File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
import java.util.*;
class Solution {
    //방향
    static int[] dy = {1, 0, -1};
    static int[] dx = {0, 1, -1};
    //2차원배열
    static int[][] temp;
    public List<Integer> solution(int n) {
        temp = new int[n + 1][n + 1];
        move(0, n, 1, 0, 0);
        List<Integer> ret = new ArrayList<>();
        for(int[] arr : temp){
            for(int a : arr){
                if(a == 0) continue;
                ret.add(a);
            }
        }
        return ret;
    }
    private void move(int d, int n, int startNum, int y, int x){
        if(n == 0) return;
        int cy = y;
        int cx = x;
        int number = startNum;
        for(int i = 0; i < n; i++){
            int ny = cy + dy[d % 3];
            int nx = cx + dx[d % 3];
            temp[cy][cx] = number;
            number++;
            if(i == n - 1) break;
            cy = ny;
            cx = nx;
        }
        // 방향 더하기, 크기 -1, 숫자는 더한거 그대로 가지고 하기, 시작점 방향 바꾼거 더하기
        move(d + 1, n - 1, number, cy + dy[(d + 1) % 3], cx + dx[(d+1) % 3]);
    }
}
/*
1.더하는 값은 고정, 방향과 크기는 변화
*/

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    tri = [[0]*(i+1) for i in range(n)]
    dy, dx = [1, 0, -1], [0, 1, -1]
    y, x, d, num = 0, 0, 0, 1
    
    while True:
        tri[y][x] = num
        num += 1
        ny, nx = y + dy[d], x + dx[d]
        if 0 <= ny < n and 0 <= nx < n and tri[ny][nx] == 0:
            y, x = ny, nx
        else:
            d = (d+1) % 3
            ny, nx = y + dy[d], x + dx[d]
            
            if 0 <= ny < n and 0 <= nx < n and tri[ny][nx] == 0:
                y, x = ny, nx
            else:
                break
        
    return [num for arr in tri for num in arr]



"""
시작좌표 = (0, 0), 시작숫자 = 1
 0      1        2
y + 1, x + 1, y - 1 x - 1
1. 다음을 반복
    1-1. 자리에 숫자을 채운다
    1-2. 경계에 닿으면 방향을 바꿔준다
    1-3. 다음 자리에 숫자가 있으면 방향을 바꿔준다
    1-4. 두번 예외인 경우이면 반복을 그만한다.
2. 맨위에서부터 저장된 숫자를 가져와서 다시 넣어준다.
"""